### ISSUE TITLE
Have you set this up to support any kind of batched "single pass rendering"?

### ISSUE THREAD

**Issue by tdbe** [janhsimon/xrvk#1]
Hey, nice project. Can you pls elaborate a little bit on how you render the geometry? Do you use the classic old multipass way of one swapchain per eye and render all geometry twice? Or do you use texture array / image array / target array, and batch-render each object to both indexes? Or do you use a double-wide swapchain with both eyes, with two different imageRects? 🙂🙃

I'm new to openxr sdk, and am searching around for the sake of my sanity and reference/verification, if there are any boilerplates out there that have actually done some kind of single pass rendering with openxr. (nothing comes up in search beyond unity and unreal)

**Comment by janhsimon** [comment]
Hi @tdbe, great question, I am planning to add much better documentation for how this and other core aspects work in this project soon.

So I asked myself the same question when I started out: What's the most efficient way to render to the headset lenses and the mirror view on a desktop with Vulkan? In fact a major reason for why I decided to make this public resource was to help others get on the right track with OpenXR and Vulkan because there are just so many questions to answer and choices to make like this one.

I did some research and asked in a Vulkan discord server, where experts pointed me to the `VK_KHR_multiview` [extension](https://registry.khronos.org/vulkan/specs/1.3-extensions/man/html/VK_KHR_multiview.html) (core in Vulkan 1.1 and later). This is the official description:

> This extension has the same goal as the OpenGL ES GL_OVR_multiview extension. Multiview is a rendering technique originally designed for VR where it is more efficient to record a single set of commands to be executed with slightly different behavior for each “view”.
> 
> It includes a concise way to declare a render pass with multiple views, and gives implementations freedom to render the views in the most efficient way possible. This is done with a multiview configuration specified during [render pass](https://registry.khronos.org/vulkan/specs/1.3-extensions/html/vkspec.html#renderpass) creation with the [VkRenderPassMultiviewCreateInfo](https://registry.khronos.org/vulkan/specs/1.3-extensions/man/html/VkRenderPassMultiviewCreateInfo.html) passed into [VkRenderPassCreateInfo](https://registry.khronos.org/vulkan/specs/1.3-extensions/man/html/VkRenderPassCreateInfo.html)::pNext.
> 
> This extension enables the use of the [SPV_KHR_multiview](https://htmlpreview.github.io/?https://github.com/KhronosGroup/SPIRV-Registry/blob/master/extensions/KHR/SPV_KHR_multiview.html) shader extension, which adds a new ViewIndex built-in type that allows shaders to control what to do for each view. If using GLSL there is also the [GL_EXT_multiview](https://github.com/KhronosGroup/GLSL/blob/master/extensions/ext/GL_EXT_multiview.txt) extension that introduces a highp int gl_ViewIndex; built-in variable for vertex, tessellation, geometry, and fragment shaders.

It's tailor-made for XR applications, super fast and even made the code leaner, which is why I used it in this project.

In a nutshell, my Headset class sets up OpenXR to use a single swapchain with multiview images, instead of the classic setup with one swapchain per eye. For each frame, I acquire the multiview swapchain image from OpenXR, then, in the Renderer class, I render my scene into that (using a single pass), both eyes at the same time (the uniform buffer holds the two view/projection matrices, and the vertex shader uses `gl_ViewIndex` to pick the correct eye transformation). Then the mirror view comes in and blits from the headset image to its window surface.

All of this is done using a single queue and command buffer as well, which makes correct synchronization trivial.

Hope this helps!

**Comment by tdbe** [comment]
Yes! Beautiful. Helps a lot! Indeed there needs to be more open discussion on how to actually do efficient vr rendering; the builtin khronos hello_xr code is straight up not how you're supposed to render xr games.

The high level nutshell description you gave is super valuable and you should def put in your readme, because it also helps know how to read the code and understand the strategy when you're new to both openXR sdk and vulkan. 👍

I will try your multiview approach

**Comment by tdbe** [comment]
> For each frame, I acquire the multiview swapchain image from OpenXR, then, in the Renderer class, I render my scene into that (using a single pass), both eyes at the same time

Wait I'm confused. 
You say you have one swapchain. With 2 swapchain images? Shouldn't it be one swapchain image with 2 layers (multiview)?
You renderer.render(swapchainImageIndex) in a while loop, not in a per-eye for loop. The swapchain image index comes from headset.beginFrame() - xrAcquireSwapchainImage. This strongly implies one pass / **one swapchain image** / one frame / one call of beginFrame per game loop.
Then in the render function you get `headset->getRenderTarget(swapchainImageIndex)` ONCE

BUT in Headset.cpp constructor, it appears you have **an array** of `swapchainImages`, and you create one renderTarget per swapchain image `renderTarget = new RenderTarget(device, image, depthImageView, eyeResolution, colorFormat, renderPass, 2u); ` sharing a renderPass. 

Is the swapchain image array and rendertarget array, of size 1? Why are they arrays then? And you do refer to them as "swapchain imageS".

Or do you somehow have one xr swapchain with 2 vk images? and then to those 2 vkimages is actually bound a multiview image from the vk physical device that is set to use multiview? Confused how the data looks.

------------------------

[EDIT]
I can see you set the `layerCount = 2u` as well for the depth image. But I can't tell where this is set up for the regular, non-depth image??

------------------------
[EDIT[

Oh I get it:
`swapchainCreateInfo.arraySize = static_cast<uint32_t>(eyeCount);` ArraySize is the layers in the image, and you set 2: one per eye `static_cast<uint32_t>(eyeCount);`. So this is for the multiview image.

Then you create a swapchain with that info, and the `xrEnumerateSwapchainImages` magically comes up with 2 images, even though it's one damn image with 2 layers. That's what I was missing. Wow these obscure ass APIs (both vulkan and openxr), where the only state is in your mind, or hidden behind unconnected layers of magic.

That's why your `swapchainImageCount` is 2.

Then after a bit of cast magic (which I assume is to create two VkImages, that alias the same memory that the image array is using),
You create the render targets which are fed one VKImage per render target. And each VKImage is actually the multiview layer.

I wrote a lot, sorry, don't expect you to read it al, might help others though. I think I got it right 😛👍

**Comment by janhsimon** [comment]
And that frustration at OpenXR and Vulkan is exactly why I created this project. 😊

But yes, you got it!
